{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7dbfdc14",
   "metadata": {},
   "source": [
    "# 合并测序数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59d0e54e",
   "metadata": {},
   "outputs": [],
   "source": [
    "BASE_DIR=\"/Users/john/git/UTR/data/prompt_sev\"\n",
    "fwd_fastq=\"${BASE_DIR}/15-8-25-1_raw_1.fq.gz\"\n",
    "rev_fastq=\"${BASE_DIR}/15-8-25-2_raw_1.fq.gz\"\n",
    "merge_fasta=\"${BASE_DIR}/sev240222_v1.fasta\"\n",
    "merge_fastq=\"${BASE_DIR}/sev240222_v1.fastq\"\n",
    "\n",
    "# 1. 利用pandaseq处理测序数据\n",
    "pandaseq -f ${fwd_fastq} -r ${rev_fastq} -w ${merge_fasta} -g log.txt\n",
    "\n",
    "# 2. 利用pear和seqkit处理测序数据\n",
    "pear -f ${fwd_fastq} -r ${rev_fastq} -p 0.001 -j 20 -n 110 -o ${merge_fastq}\n",
    "\n",
    "# 用于pear输出的fastq格式，需要seqkit进一步处理测序数据\n",
    "seqkit fq2fa ${merge_fastq} -o ${merge_fasta}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03123600",
   "metadata": {},
   "source": [
    "# 处理DNA数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f77685b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import tqdm\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "from collections import Counter\n",
    "\n",
    "# DNA_regexp = re.compile(r\"ggtgcctgaaactag([a-z]+)atg([a-z]{9,21})ggtggttcgggcggt\",re.I) # 0610\n",
    "DNA_regexp = re.compile(r\"gaccaactag([a-z]+)atg([a-z]{15})ggtggttcgg\",re.I) # 11\n",
    "\n",
    "\n",
    "\n",
    "def record_dna(match, m,d):\n",
    "    target_seq = match.group(1)\n",
    "    barcode = match.group(2)\n",
    "    d[barcode]=d.get(barcode,0)+1\n",
    "    counter = m.get(barcode,Counter())\n",
    "    counter.update([target_seq])\n",
    "    m[barcode]=counter\n",
    "\n",
    "\n",
    "def gendna_abundance(fn):\n",
    "    d, m= {}, {}\n",
    "    reverse, forward = 0,0\n",
    "    for i, v in enumerate(tqdm.tqdm(SeqIO.parse(fn, format=\"fasta\"))):\n",
    "#         if i>200000:\n",
    "#             break\n",
    "        seq = str(v.seq)\n",
    "        for match in re.finditer(DNA_regexp, seq):\n",
    "            forward +=1\n",
    "            record_dna(match,m,d)\n",
    "        seq_reverse = str(v.seq.reverse_complement())\n",
    "        for match in re.finditer(DNA_regexp, seq_reverse):\n",
    "            reverse +=1\n",
    "            record_dna(match,m,d)\n",
    "    barcodes,seqs,counts = [],[],[]\n",
    "    for b,c in m.items():\n",
    "        for s,n in c.items():\n",
    "            barcodes.append(b)\n",
    "            seqs.append(s)\n",
    "            counts.append(n)\n",
    "    abundance_count_df = pd.DataFrame(data={\"barcode\":barcodes, \"dna_seq\":seqs,\"plasmid_counts\":counts})\n",
    "    print(len(abundance_count_df))\n",
    "    base,fname = os.path.dirname(fn),os.path.basename(fn)\n",
    "    name = fname[:fname.find(\".\")]\n",
    "    abundance_count_df.to_csv(os.path.join(base,name+\".csv\"),index=False)\n",
    "    print(f\"found {forward} sequences forward, {reverse} reverse\")\n",
    "    print(f\"Converting {forward+reverse} sequences to dataframe and saving\")\n",
    "    return abundance_count_df\n",
    "    \n",
    "gendna_abundance(fn=\"/data/home/jinyalong/data/SeV_231101/DNA.fa\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b53c12ff",
   "metadata": {},
   "source": [
    "# 处理RNA数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fffca399",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import tqdm\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "from collections import Counter\n",
    "\n",
    "# RNA_regexp = re.compile(r\"GCAGAGTACAT([a-z]+)atg([a-z]{9,21})ggtggttcgg\",re.I)\n",
    "RNA_regexp = re.compile(r\"^([a-z]+)atg([a-z]{15})ggtggttcgg\",re.I)\n",
    "\n",
    "def record_rna(match, m,d):\n",
    "    target_seq = match.group(1)\n",
    "    barcode = match.group(2)\n",
    "    d[barcode]=d.get(barcode,0)+1\n",
    "    counter = m.get(barcode,Counter())\n",
    "    counter.update([target_seq])\n",
    "    m[barcode]=counter\n",
    "    \n",
    "\n",
    "def genrna_abundance(fn):\n",
    "    m, d={}, {}\n",
    "    reverse, forward = 0,0\n",
    "    for i, v in enumerate(tqdm.tqdm(SeqIO.parse(fn, format=\"fasta\"))):\n",
    "#         if i>200000:\n",
    "#             break\n",
    "        seq = str(v.seq)\n",
    "        for match in re.finditer(RNA_regexp, seq):\n",
    "            forward +=1\n",
    "            record_rna(match,m,d)\n",
    "        seq_reverse = str(v.seq.reverse_complement())\n",
    "        for match in re.finditer(RNA_regexp, seq_reverse):\n",
    "            reverse +=1\n",
    "            record_rna(match,m,d)\n",
    "    barcodes,seqs,counts = [],[],[]\n",
    "    for b,c in m.items():\n",
    "        for s,n in c.items():\n",
    "            barcodes.append(b)\n",
    "            seqs.append(s)\n",
    "            counts.append(n)\n",
    "    abundance_count_df = pd.DataFrame(data={\"barcode\":barcodes, \"rna_seq\":seqs,\"rna_counts\":counts})\n",
    "    print(len(abundance_count_df))\n",
    "    base,fname = os.path.dirname(fn),os.path.basename(fn)\n",
    "    name = fname[:fname.find(\".\")]\n",
    "    print(f\"found {forward} sequences forward, {reverse} reverse\")\n",
    "    print(f\"Converting {forward+reverse} sequences to dataframe and saving\")\n",
    "    abundance_count_df.to_csv(os.path.join(base,name+\".csv\"),index=False)\n",
    "    return abundance_count_df\n",
    "\n",
    "# genrna_abundance(fn=\"/data/home/jinyalong/data/SeV_240618/pl3-2-1.fa\")\n",
    "# genrna_abundance(fn=\"/data/home/jinyalong/data/SeV_240618/pl3-2-2.fa\")\n",
    "genrna_abundance(fn=\"/data/home/jinyalong/data/SeV_231101/RNA1.fa\")\n",
    "genrna_abundance(fn=\"/data/home/jinyalong/data/SeV_231101/RNA2.fa\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18fc0e64",
   "metadata": {},
   "source": [
    "# 检查DNA和RNA数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c1f75d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# dna_seq = pd.read_csv(\"/data/home/jinyalong/data/SeV_240530/DNA2.csv\")\n",
    "# rna_seq = pd.read_csv(\"/data/home/jinyalong/data/SeV_240530/pl3-2-1.csv\")\n",
    "\n",
    "dna_seq = pd.read_csv(\"/data/home/jinyalong/data/SeV_231101/DNA.csv\")\n",
    "rna_seq = pd.read_csv(\"/data/home/jinyalong/data/SeV_231101/RNA1.csv\")\n",
    "\n",
    "original_db = set([ v.upper() for v in pd.read_csv(\"mutation_db_118.csv\")[\"seq\"].values])\n",
    "dna_seq_db = dna_seq[dna_seq.dna_seq.isin(original_db)]\n",
    "dna_seq_dedup = set(dna_seq[\"dna_seq\"].drop_duplicates().values)\n",
    "dna_seq_dedup_db = original_db & dna_seq_dedup\n",
    "dna_seq_barcode_dedup = dna_seq[\"barcode\"].drop_duplicates().values\n",
    "dna_seq_barcode_dedup_db = dna_seq_db[\"barcode\"].drop_duplicates().values\n",
    "print(f'''Read DNAs {dna_seq[\"plasmid_counts\"].sum()} reads found, {dna_seq_db[\"plasmid_counts\"].sum()} in db; \n",
    "      {len(dna_seq_dedup)} types, {len(dna_seq_dedup_db)} in db; \n",
    "      {len(dna_seq_barcode_dedup)} barcodes, {len(dna_seq_barcode_dedup_db)} in db''')\n",
    "\n",
    "\n",
    "rna_seq_dedup = rna_seq.groupby(['rna_seq'])[\"rna_counts\"].sum().reset_index(name='rna_counts')\n",
    "rna_seq_barcode_dedup = rna_seq.groupby(['barcode'])[\"rna_counts\"].sum().reset_index(name='barcode_counts')\n",
    "print(f'''Read RNAs {rna_seq[\"rna_counts\"].sum()} reads found, \n",
    "      {len(rna_seq_dedup)} types, {len(rna_seq_barcode_dedup)} barcodes''')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e01345ae",
   "metadata": {},
   "source": [
    "# 清洗DNA和RNA数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cd2c329",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df=pd.read_csv(\"/data/home/jinyalong/data/SeV_240530/DNA2.csv\")\n",
    "# df = df[df[\"plasmid_counts\"]>5]\n",
    "barcodes, seqs, counts = [],[],[]\n",
    "bar1,barm, hit=0,0,0\n",
    "for k, g in df.groupby([\"barcode\"]):\n",
    "    if g.shape[0]==1:\n",
    "        bar1+=1\n",
    "        barcodes.append(g.iloc[0].barcode)\n",
    "        seqs.append(g.iloc[0].dna_seq)\n",
    "        counts.append(g.iloc[0].plasmid_counts)\n",
    "    else:\n",
    "        barm+=1\n",
    "        lst = sorted(g[\"plasmid_counts\"].values)\n",
    "        gdf = g.sort_values(by=['plasmid_counts'],ascending=False)\n",
    "        top1, top2 = gdf.iloc[0].plasmid_counts,gdf.iloc[1].plasmid_counts\n",
    "        if top1>10*top2:\n",
    "            hit += 1\n",
    "            barcodes.append(gdf.iloc[0].barcode)\n",
    "            seqs.append(gdf.iloc[0].dna_seq)\n",
    "            counts.append(gdf.iloc[0].plasmid_counts)\n",
    "print(bar1, barm, hit)\n",
    "cdf = pd.DataFrame({\"barcode\":barcodes,\"dna_seq\":seqs,\"plasmid_counts\":counts})\n",
    "cdf.to_csv(\"/data/home/jinyalong/data/SeV_240530/DNA2_clean.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9113eca6",
   "metadata": {},
   "source": [
    "# 合并DNA和RNA数据"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6347c8a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna1=pd.read_csv(\"/data/home/jinyalong/data/SeV_240530/pl3-1-1.csv\")\n",
    "rna2=pd.read_csv(\"/data/home/jinyalong/data/SeV_240530/pl3-1-2.csv\")\n",
    "original_db = set([ v.upper() for v in pd.read_csv(\"./data/mutation_db_118.csv\")[\"seq\"].values])\n",
    "ddf = pd.read_csv(\"/data/home/jinyalong/data/SeV_240530/DNA2_clean.csv\")\n",
    "ddf = ddf[ddf[\"dna_seq\"].isin(original_db)]\n",
    "rna1 = rna1.groupby(['barcode'])[\"rna_counts\"].sum().reset_index(name='rna1')\n",
    "rna2 = rna2.groupby(['barcode'])[\"rna_counts\"].sum().reset_index(name='rna2')\n",
    "mdf = pd.merge(ddf,rna1,on=[\"barcode\"], how=\"left\")\n",
    "mdf = pd.merge(mdf,rna2,on=[\"barcode\"], how=\"left\")\n",
    "mdf.fillna(0, inplace=True)\n",
    "mdf = mdf[(mdf[\"rna1\"]>0)|(mdf[\"rna2\"]>0)]\n",
    "\n",
    "def std(r1,r2):\n",
    "    if r1 and r2:\n",
    "        return max(r1/r2,r2/r1)\n",
    "    else:\n",
    "        return 100\n",
    "    \n",
    "mdf[\"std\"] = mdf.apply(lambda x:std(x[\"rna1\"],x[\"rna2\"]),axis=1)\n",
    "mdf[\"score1\"] = mdf[\"rna1\"]/mdf[\"plasmid_counts\"]\n",
    "mdf[\"score2\"] = mdf[\"rna2\"]/mdf[\"plasmid_counts\"]\n",
    "mean = (mdf[\"score1\"].mean()+mdf[\"score2\"].mean())/2\n",
    "mdf[\"sstd\"] = mdf.apply(lambda x:abs(x[\"score1\"]-x[\"score2\"])/mean,axis=1)\n",
    "cdf = mdf\n",
    "# cdf = mdf[(mdf[\"sstd\"]<0.2)]\n",
    "cdf = mdf[(mdf[\"sstd\"]<0.2)]\n",
    "# fdf = cdf.groupby(\"dna_seq\")[\"plasmid_counts\",\"rna1\",\"rna2\"].sum().reset_index()\n",
    "fdf = cdf\n",
    "fdf[\"seq\"] = fdf.apply(lambda x:x[\"dna_seq\"]+\"ATG\"+x[\"barcode\"],axis=1)\n",
    "fdf[\"rna_counts\"]=(fdf[\"rna1\"]+fdf[\"rna2\"])/2\n",
    "# mrna=cdf.groupby(\"barcode\")[\"rna2\"].sum().to_dict()\n",
    "# cdf[\"rna_counts\"] = cdf[\"barcode\"].map(mrna)\n",
    "fdf.to_csv(\"./data/DNA_RNA_DAY5_BARCODE.csv\",index=False)\n",
    "fdf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09ad723b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "import Levenshtein\n",
    "import numpy as np\n",
    "original_seq = 'atcccgggtgaggcatcccaccatcctcagtcacagagagacccaatctaccatcagcatcagccagtaaagattaagaaaaacttagggtgaaagaaatttcacctaacacggcgca'.upper()\n",
    "dna_seq = pd.read_csv(\"/data/home/jinyalong/data/SeV_240530/DNA2.csv\")\n",
    "rna_seq = pd.read_csv(\"/data/home/jinyalong/data/SeV_240530/pl3-1-1.csv\")\n",
    "original_db = set([ v.upper() for v in pd.read_csv(\"mutation_db_118.csv\")[\"seq\"].values])\n",
    "dna_seq_db = dna_seq[dna_seq.dna_seq.isin(original_db)]\n",
    "rna_barcode = rna_seq.groupby(['barcode'])[\"rna_counts\"].sum().reset_index(name='rna_counts')\n",
    "dna_rna_db = dna_seq_db[dna_seq_db.barcode.isin(rna_barcode.barcode)]\n",
    "barcode_1vn_threshold = 10  # 对barcode一对多的情况，只有最多的比第二多出barcode_1vn_threshold倍才可用于计数\n",
    "raw_path = \"raw_data.csv\"\n",
    "score_path = \"score_file.csv\"\n",
    "errors = []\n",
    "errors_group_count = 0\n",
    "dna_rna_db.rename(columns={\"dna_seq\":\"seq\"},inplace=True)\n",
    "dna_seq_db=dna_seq_db.copy()\n",
    "dna_seq_db.rename(columns={\"dna_seq\":\"seq\"},inplace=True)\n",
    "\n",
    "with open(raw_path, 'w') as wfh:\n",
    "    for g, group in tqdm(dna_rna_db.groupby('barcode')):\n",
    "        group['dna_types'] = len(set(group['seq']))\n",
    "        group = group[['barcode', 'dna_types', 'seq']]\n",
    "        reads = group.groupby('seq').size().reset_index(name='dna_reads')\n",
    "        group = group.merge(reads).drop_duplicates('seq')\n",
    "        group = group.merge(rna_barcode)\n",
    "        group = group[group.dna_reads > (group.dna_reads.max() / barcode_1vn_threshold)]\n",
    "        if len(group) > 1:\n",
    "            errors.append(group)\n",
    "            errors_group_count += 1\n",
    "        else:\n",
    "            group.to_csv(wfh, index=False, header=None, sep='\\t')\n",
    "\n",
    "dna_rna_db_merge = pd.read_csv(raw_path, sep='\\t', header=None)\n",
    "dna_rna_db_merge.columns = ['barcode', 'barcode_matched_dna_types', 'seq', 'plasmid_counts', 'rna_counts']\n",
    "dna_rna_db_merge = dna_rna_db_merge.sort_values(by=['rna_counts', 'plasmid_counts', 'barcode_matched_dna_types'], ascending=False)\n",
    "dna_rna_db_merge.to_csv(raw_path, sep='\\t', index=False)\n",
    "dna_rna_db_merge = dna_rna_db_merge.groupby('seq').sum().reset_index(drop=False)[['seq', 'plasmid_counts', 'rna_counts']]\n",
    "dna_rna_db_merge['length'] = dna_rna_db_merge.seq.apply(len)\n",
    "plasmid_counts_dict = dna_seq_db.groupby('seq')[\"plasmid_counts\"].sum().to_dict()\n",
    "dna_rna_db_merge['plasmid_counts'] = dna_rna_db_merge.seq.apply(lambda x: plasmid_counts_dict[x])\n",
    "dna_rna_db_merge['plasmid_freq'] = dna_rna_db_merge.plasmid_counts / dna_rna_db_merge.plasmid_counts.sum()\n",
    "dna_rna_db_merge['rna_freq'] = dna_rna_db_merge.rna_counts / dna_rna_db_merge.rna_counts.sum()\n",
    "dna_rna_db_merge['score'] = np.log(dna_rna_db_merge.rna_freq / dna_rna_db_merge.plasmid_freq)\n",
    "dna_rna_db_merge['distance'] = dna_rna_db_merge.seq.apply(lambda x: Levenshtein.distance(original_seq, x))\n",
    "distance_bins = np.arange(0, dna_rna_db_merge.distance.max() // 5 * 4, dna_rna_db_merge.distance.max() // 5)\n",
    "distance_bins = np.append(distance_bins, np.inf)\n",
    "dna_rna_db_merge['distance_cat'] = pd.cut(dna_rna_db_merge.distance, bins=distance_bins, labels=list(range(1, len(distance_bins))), right=False)\n",
    "dna_rna_db_merge = dna_rna_db_merge[['seq', 'plasmid_counts', 'length', 'plasmid_freq', 'rna_counts', 'rna_freq', 'score', 'distance', 'distance_cat']]\n",
    "\n",
    "dna_rna_db_merge.to_csv(score_path)\n",
    "errors = pd.concat(errors)\n",
    "print(f'finally used {dna_rna_db_merge.plasmid_counts.sum()} DNA reads, {dna_rna_db_merge.rna_counts.sum()} RNA reads, {len(dna_rna_db_merge)} different sequences')\n",
    "print(f'{errors_group_count} types of RNA cannot be matched to DNA sequence due to 1vn problem, {errors.rna_counts.sum()} reads in summary')\n",
    "print('wild type exist', str(bool(len(dna_rna_db_merge[dna_rna_db_merge.seq == original_seq]))))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
