{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8ffedb50",
   "metadata": {},
   "source": [
    "# MPRA training (for analyses)\n",
    "\n",
    "This notebook process the MPRA results to generate the table used to produce the analyses reported in the CODA paper. A slightly different table is used to optimize Malinois weights, which can be created using the notebook\"CODA_preprocess_MPRA_validation\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e106fe8c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/sq/pxxh686s20scbcsrvq340czj6yvqp0/T/ipykernel_44951/2667858688.py:6: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display\n",
      "  from IPython.core.display import display, HTML\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>.container { width:98% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gzip\n",
    "\n",
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:98% !important; }</style>\"))\n",
    "\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98ab1372",
   "metadata": {},
   "source": [
    "### Fetch the MPRA results\n",
    "\n",
    "The dataset consists of oligo libraries (OL) from the MPRA projects UKBB (OL27-33), GTEx (OL41,42), and CRE (OL15)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bf17a31b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching files:\n",
      "\n",
      "OL42_HepG2_20211010.out\n",
      "OL28_SKNSH_20210428.out\n",
      "OL41_SKNSH_20211010.out\n",
      "OL42_SKNSH_20211010.out\n",
      "OL28_HEPG2_20210428.out\n",
      "OL27_K562_20210428.out\n",
      "OL33_HEPG2_20210428.out\n",
      "OL15_HEPG2_Neon_20200904.out\n",
      "OL30_SKNSH_20210428.out\n",
      "OL33_SKNSH_20210428.out\n",
      "OL41B_HepG2_20211010.out\n",
      "OL30_HEPG2_20210428.out\n",
      "OL32_SKNSH_20210428.out\n",
      "OL33_K562_20210428.out\n",
      "OL15_K562_20200904.out\n",
      "OL31_HEPG2_20210428.out\n",
      "OL15_SKNSH_20200904.out\n",
      "OL32_HEPG2_20210428.out\n",
      "OL32_K562_20210428.out\n",
      "OL31_SKNSH_20210428.out\n",
      "OL27_SKNSH_20210428.out\n",
      "OL28_K562_20210428.out\n",
      "OL31_K562_20210428.out\n",
      "OL29_HEPG2_20210428.out\n",
      "OL27_HEPG2_20210428.out\n",
      "OL29_K562_20210428.out\n",
      "OL30_K562_20210428.out\n",
      "OL29_SKNSH_20210428.out\n",
      "OL41-42_K562_20211010.out\n"
     ]
    }
   ],
   "source": [
    "rootdir = '/training'\n",
    "\n",
    "cell_types_lower = ['k562', 'hepg2', 'sknsh']\n",
    "cell_types = ['K562', 'HepG2', 'SKNSH']\n",
    "std_cell_names = dict(zip(cell_types_lower, cell_types))\n",
    "\n",
    "UKBB_libraries = [f'OL{n}' for n in range(27, 34)]\n",
    "GTEx_libraries = ['OL41', 'OL42', 'OL41-42', 'OL41B']\n",
    "CREs_libraries = ['OL15']\n",
    "\n",
    "data_project_dict = {OL: 'UKBB' for OL in UKBB_libraries}\n",
    "data_project_dict.update({OL: 'GTEx' for OL in GTEx_libraries})\n",
    "data_project_dict.update({OL: 'CRE' for OL in CREs_libraries})\n",
    "\n",
    "print('Fetching files:\\n')\n",
    "in_df = []\n",
    "for subdir, dirs, files in os.walk(rootdir):\n",
    "    for file in files:\n",
    "        filepath = subdir + os.sep + file\n",
    "        if filepath.endswith(\".out\") and ('Counts' not in filepath):\n",
    "            library_num, cell_type = file.split('_')[:2]\n",
    "            if cell_type.lower() in cell_types_lower:\n",
    "                df_temp = pd.read_csv(filepath, sep='\\t', low_memory=False)\n",
    "                df_temp['OL'] = library_num\n",
    "                df_temp['cell_type'] = std_cell_names[cell_type.lower()]\n",
    "                df_temp['data_project'] = data_project_dict[library_num]\n",
    "                in_df.append(df_temp)\n",
    "                print(file)\n",
    "\n",
    "in_columns = ['ID', 'chr', 'project', 'ctrl_exp', 'Ctrl.Mean',\n",
    "       'Exp.Mean', 'log2FoldChange', 'lfcSE', 'OL',\n",
    "       'cell_type', 'data_project']\n",
    "\n",
    "in_df = pd.concat(in_df)[in_columns]\n",
    "\n",
    "row_filter = in_df['chr'].isna()\n",
    "chrs = []\n",
    "for ID in in_df[row_filter]['ID'].tolist():\n",
    "    if 'sample' in ID:\n",
    "        chrs.append('synth')\n",
    "    else:\n",
    "        chrs.append(ID.split(':')[0])\n",
    "        \n",
    "in_df.loc[row_filter, 'chr'] = chrs\n",
    "\n",
    "in_df.loc[in_df['project'].isna(), 'project'] = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e053ae69",
   "metadata": {},
   "source": [
    "### Filter oligos with no RNA count\n",
    "\n",
    "Before aggregating across libraries, we fitler out sequences with no RNA count.\n",
    "\n",
    "Note: In order to reproduce the table used to optimize Malinois weights one would need to also filter out sequences with an oligo count less than 20."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "720191b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "oligo_count_cutoff = 0\n",
    "RNA_count_cutoff = 0\n",
    "\n",
    "quality_filter = (in_df['Ctrl.Mean'] >= oligo_count_cutoff) & \\\n",
    "                         (in_df['Exp.Mean'] > RNA_count_cutoff)\n",
    "\n",
    "in_df = in_df[quality_filter].reset_index(drop=True) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be910017",
   "metadata": {},
   "source": [
    "### Get the nucleotide sequence\n",
    "\n",
    "We load the nuecleotide sequence information from each corresponding fasta file and assign to each oligo ID its corresponding sequence. It is worth noting that there might be very rare edge cases in which the same oligo ID might differ by one position between the UKBB and GTEx databases. To our knowledge, there is only one case. Nonetheless, fetching the sequence from the corresponding OL fasta file (as opposed to just creating a set of ID-to_sequence pairs) ensures we get the correct measured sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e07d614",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing fasta files:\n",
      "\n",
      "OL41_reference.fasta.gz\n",
      "OL32_reference.fasta.gz\n",
      "OL31_reference.fasta.gz\n",
      "OL42_reference.fasta.gz\n",
      "OL27_reference.fasta.gz\n",
      "OL29_reference.fasta.gz\n",
      "OL15_reference.fasta.gz\n",
      "OL28_reference.fasta.gz\n",
      "OL30_reference.fasta.gz\n",
      "OL33_reference.fasta.gz\n"
     ]
    }
   ],
   "source": [
    "rootdir = '/Fastas'\n",
    "\n",
    "ignore_libraries = ['OL46']\n",
    "\n",
    "fasta_dict = {'UKBB': {}, 'GTEx': {}, 'CRE': {}}\n",
    "print('Processing fasta files:\\n')\n",
    "for subdir, dirs, files in os.walk(rootdir):\n",
    "    for file in files:\n",
    "        filepath = subdir + os.sep + file\n",
    "        if filepath.endswith(\".gz\") and ('fasta' in filepath):\n",
    "            library_num = file.split('_')[0]\n",
    "            if library_num not in ignore_libraries:\n",
    "                print(file)\n",
    "                project = data_project_dict[library_num]\n",
    "                with gzip.open(filepath, 'rt') as f:\n",
    "                    for line_str in f:\n",
    "                        if line_str[0] == '>':\n",
    "                            oligo_id = line_str.lstrip('>').rstrip('\\n')\n",
    "                            fasta_dict[project][oligo_id] = ''\n",
    "                        else:\n",
    "                            fasta_dict[project][oligo_id] += line_str.rstrip('\\n')\n",
    "\n",
    "for project in fasta_dict.keys():\n",
    "    project_fasta_dict = fasta_dict[project]\n",
    "    compressed_IDs = [ID for ID in project_fasta_dict.keys() if ';' in ID]\n",
    "    for compressed_ID in compressed_IDs:\n",
    "        stripped_ID = compressed_ID.lstrip('(').rstrip(')')\n",
    "        for single_ID in stripped_ID.split(';'):\n",
    "            project_fasta_dict[single_ID] = project_fasta_dict[compressed_ID]\n",
    "        \n",
    "in_df['sequence'] = in_df.apply(lambda x: fasta_dict[x['data_project']][x['ID']], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57b0c5e3",
   "metadata": {},
   "source": [
    "### Grouping results for oligos shared across libraries in a data project\n",
    "\n",
    "This step is needed only for sequences in the UKBB oligo libraries, since these libraries share some sequences. However it does hurt to do it for the libraries in GTEx and CRE. For practical purposes, we average the log2(Fold-Change) readout of sequences shared across libraries in order to have a single sequence-to-activity pair for each sequence. For measures that can be inform us of empirical noise (such as oligo count, RNA count, and lof2FC standard error), we retrieve the \"worst\" value (min, min, max, respectively) across libraries.\n",
    "\n",
    "Note: For legacy reasons, we do not average shared sequences across projects since the generation of these projects was asynchronous. However, the consistency of the MPRA results probably allows for averaging across different projects as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d5f7f3db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Grouping sequences across cell types and across libraries in UKBB\n",
      "Grouping sequences across cell types and across libraries in GTEx\n",
      "Grouping sequences across cell types and across libraries in CRE\n"
     ]
    }
   ],
   "source": [
    "grouped_project_dfs = []\n",
    "for data_project in ['UKBB', 'GTEx', 'CRE']:\n",
    "    print(f'Grouping sequences across cell types and across libraries in {data_project}')\n",
    "    project_df = in_df[in_df['data_project'] == data_project].reset_index(drop=True)\n",
    "    groups = project_df.groupby(by=['sequence', 'cell_type']).agg(\n",
    "                log2FC_mean = ('log2FoldChange', 'mean'),\n",
    "                chromosome = ('chr', 'first'),\n",
    "                OL = ('OL', set),\n",
    "                classes = ('project', set),\n",
    "                IDs = ('ID', set),\n",
    "                exp_mean = ('Exp.Mean', 'min'),\n",
    "                ctrl_mean = ('Ctrl.Mean', 'min'),\n",
    "                lfcSE = ('lfcSE', 'max'))\n",
    "    grouped_project_df = groups.unstack(level=-1)\n",
    "    grouped_project_df.columns = ['_'.join(col).rstrip('_') for col in grouped_project_df.columns.values]\n",
    "    grouped_project_df.reset_index(inplace=True)\n",
    "    grouped_project_df['data_project'] = data_project\n",
    "    grouped_project_dfs.append(grouped_project_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e1ab18e",
   "metadata": {},
   "source": [
    "### Clean the grouped columns\n",
    "\n",
    "These lines just clean up the columns generated from stacking the groups generated in the previous cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "912ab61d",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_drop_list = ['chromosome_K562', 'chromosome_SKNSH',\n",
    "               'classes_K562','classes_SKNSH',\n",
    "               'IDs_K562', 'IDs_SKNSH',\n",
    "               'OL_K562', 'OL_HepG2',\n",
    "              ]\n",
    "\n",
    "new_column_name_dict = {'sequence': 'sequence',\n",
    "                    'log2FC_mean_HepG2': 'HepG2_log2FC',\n",
    "                    'log2FC_mean_K562': 'K562_log2FC',\n",
    "                    'log2FC_mean_SKNSH': 'SKNSH_log2FC',\n",
    "                    'chromosome_HepG2': 'chr',\n",
    "                    'classes_HepG2': 'class',\n",
    "                    'IDs_HepG2': 'IDs',\n",
    "                    'OL_SKNSH': 'OL',\n",
    "                    'exp_mean_HepG2': 'exp_mean_HepG2',\n",
    "                    'exp_mean_K562': 'exp_mean_K562',\n",
    "                    'exp_mean_SKNSH': 'exp_mean_SKNSH',\n",
    "                    'ctrl_mean_HepG2': 'ctrl_mean_HepG2',\n",
    "                    'ctrl_mean_K562': 'ctrl_mean_K562',\n",
    "                    'ctrl_mean_SKNSH': 'ctrl_mean_SKNSH',\n",
    "                    'lfcSE_HepG2': 'HepG2_lfcSE',\n",
    "                    'lfcSE_K562': 'K562_lfcSE',\n",
    "                    'lfcSE_SKNSH': 'SKNSH_lfcSE',\n",
    "                    'data_project': 'data_project',\n",
    "                   }\n",
    "\n",
    "out_column_order = ['IDs', 'sequence', 'data_project', 'OL', 'class', 'chr',\n",
    "                    'K562_log2FC', 'HepG2_log2FC', 'SKNSH_log2FC',\n",
    "                    'K562_lfcSE', 'SKNSH_lfcSE', 'HepG2_lfcSE', \n",
    "#                     'ctrl_mean_K562', 'ctrl_mean_HepG2', 'ctrl_mean_SKNSH',\n",
    "#                     'exp_mean_K562', 'exp_mean_HepG2', 'exp_mean_SKNSH',\n",
    "                   ]\n",
    "\n",
    "out_df = pd.concat(grouped_project_dfs)\n",
    "\n",
    "out_df.drop(column_drop_list, axis=1, inplace=True)\n",
    "out_df.columns = [new_column_name_dict[old_name] for old_name in out_df.columns.values]\n",
    "\n",
    "null_filter = (out_df['HepG2_log2FC'].notnull()) & (out_df['K562_log2FC'].notnull()) & (out_df['SKNSH_log2FC'].notnull())\n",
    "out_df = out_df[null_filter].reset_index(drop=True)\n",
    "\n",
    "out_df['OL'] = out_df['OL'].apply(lambda x: ','.join(sorted(x)))\n",
    "out_df['class'] = out_df['class'].apply(lambda x: ','.join(sorted(x)))\n",
    "out_df['IDs'] = out_df['IDs'].apply(lambda x: ';'.join(sorted(x)))\n",
    "\n",
    "compressed_ids_filter = out_df['IDs'].str.contains(';')\n",
    "out_df.loc[compressed_ids_filter, 'IDs'] = out_df[compressed_ids_filter]['IDs'].apply(lambda x: '(' + ';'.join(x.split(';')) + ')')\n",
    "\n",
    "out_df = out_df[out_column_order]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc1a63da",
   "metadata": {},
   "source": [
    "### Drop duplicated sequences across projects\n",
    "\n",
    "In order to keep a single sequence-to-activity pair for each sequence, we drop duplicated oligos across projects. If a duplicated oligo is present in UKBB, we keep only the UKBB information. If a duplicated oligo is present in GTEx and CRE but not UKBB, we keep only the GTEx information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "52963281",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IDs</th>\n",
       "      <th>sequence</th>\n",
       "      <th>data_project</th>\n",
       "      <th>OL</th>\n",
       "      <th>class</th>\n",
       "      <th>chr</th>\n",
       "      <th>K562_log2FC</th>\n",
       "      <th>HepG2_log2FC</th>\n",
       "      <th>SKNSH_log2FC</th>\n",
       "      <th>K562_lfcSE</th>\n",
       "      <th>SKNSH_lfcSE</th>\n",
       "      <th>HepG2_lfcSE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7:89640175:A:G:A:wC</td>\n",
       "      <td>AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGAATGTCAGTT...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL29</td>\n",
       "      <td>BMI</td>\n",
       "      <td>7</td>\n",
       "      <td>0.955879</td>\n",
       "      <td>0.292685</td>\n",
       "      <td>-0.780102</td>\n",
       "      <td>0.559413</td>\n",
       "      <td>2.904046</td>\n",
       "      <td>1.148763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6:159961451:G:A:A:wC</td>\n",
       "      <td>GGTGTCAGAACGCCCCAGGGGGCTTTGAGGGCTGCTGCCCCTGGAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL30</td>\n",
       "      <td>SBP</td>\n",
       "      <td>6</td>\n",
       "      <td>2.059735</td>\n",
       "      <td>0.986049</td>\n",
       "      <td>-0.431290</td>\n",
       "      <td>0.097276</td>\n",
       "      <td>0.089045</td>\n",
       "      <td>0.087488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18:41953230:C:T:A:wC</td>\n",
       "      <td>GGTGTCACAGTAATGTGGAGGAGTAGAGGCACTCTGGCTTTTTGAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL27</td>\n",
       "      <td>Plt</td>\n",
       "      <td>18</td>\n",
       "      <td>0.256503</td>\n",
       "      <td>0.266996</td>\n",
       "      <td>-0.177992</td>\n",
       "      <td>0.088237</td>\n",
       "      <td>0.111069</td>\n",
       "      <td>0.100591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18:41953230:C:T:R:wC</td>\n",
       "      <td>GGTGTCACAGTAATGTGGAGGAGTAGAGGCACTCTGGCTTTTTGAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL27</td>\n",
       "      <td>Plt</td>\n",
       "      <td>18</td>\n",
       "      <td>0.353888</td>\n",
       "      <td>0.132264</td>\n",
       "      <td>-0.052014</td>\n",
       "      <td>0.080057</td>\n",
       "      <td>0.105239</td>\n",
       "      <td>0.100011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21:43635174:A:G:A:wC</td>\n",
       "      <td>GGTGTCAAGTGAGCGAGATGTCCCCGGCCACACAGTGAAGCACAAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL32</td>\n",
       "      <td>Age_at_Menarche</td>\n",
       "      <td>21</td>\n",
       "      <td>-0.174600</td>\n",
       "      <td>-0.422087</td>\n",
       "      <td>-0.416915</td>\n",
       "      <td>0.184045</td>\n",
       "      <td>0.168331</td>\n",
       "      <td>0.212965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798059</th>\n",
       "      <td>18:9125893:NA:NA</td>\n",
       "      <td>CAGTACTGCTGGCCCCAGAAAAGCCCCTCTCCTTATACCCTAGGCC...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>18</td>\n",
       "      <td>-0.204913</td>\n",
       "      <td>-0.156933</td>\n",
       "      <td>-0.209358</td>\n",
       "      <td>0.133052</td>\n",
       "      <td>0.185115</td>\n",
       "      <td>0.157279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798060</th>\n",
       "      <td>12:33905808:NA:NA</td>\n",
       "      <td>CAGTACCTTGTCCCCACTTCCCATTTGGCCTCTGGCAGAGGAGGAG...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>1.218233</td>\n",
       "      <td>0.613623</td>\n",
       "      <td>0.569894</td>\n",
       "      <td>0.127132</td>\n",
       "      <td>0.190639</td>\n",
       "      <td>0.167222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798061</th>\n",
       "      <td>3:128145854:NA:NA</td>\n",
       "      <td>CAGTACACCCCAGCTTCCAAAGGCCTTCTGTGACAAAGAGAGACTA...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>-0.222234</td>\n",
       "      <td>-0.338764</td>\n",
       "      <td>-0.817852</td>\n",
       "      <td>0.159002</td>\n",
       "      <td>0.238637</td>\n",
       "      <td>0.198187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798062</th>\n",
       "      <td>8:134705703:NA:NA</td>\n",
       "      <td>CAGGTTGAAGGGAATCCAAGGGCCATTTCCTTGGGAATTTCAGAAA...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "      <td>-0.194952</td>\n",
       "      <td>-0.199087</td>\n",
       "      <td>0.169847</td>\n",
       "      <td>0.131227</td>\n",
       "      <td>0.170301</td>\n",
       "      <td>0.152272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798063</th>\n",
       "      <td>2:42731164:NA:NA</td>\n",
       "      <td>TTTTTTTTTTTTTTTTGAGATGGAGTTTCCCTCTTGTCACCCAGGC...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>1.064113</td>\n",
       "      <td>0.099737</td>\n",
       "      <td>0.525217</td>\n",
       "      <td>0.118288</td>\n",
       "      <td>0.171828</td>\n",
       "      <td>0.161234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>798064 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         IDs  \\\n",
       "0        7:89640175:A:G:A:wC   \n",
       "1       6:159961451:G:A:A:wC   \n",
       "2       18:41953230:C:T:A:wC   \n",
       "3       18:41953230:C:T:R:wC   \n",
       "4       21:43635174:A:G:A:wC   \n",
       "...                      ...   \n",
       "798059      18:9125893:NA:NA   \n",
       "798060     12:33905808:NA:NA   \n",
       "798061     3:128145854:NA:NA   \n",
       "798062     8:134705703:NA:NA   \n",
       "798063      2:42731164:NA:NA   \n",
       "\n",
       "                                                 sequence data_project    OL  \\\n",
       "0       AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGAATGTCAGTT...         UKBB  OL29   \n",
       "1       GGTGTCAGAACGCCCCAGGGGGCTTTGAGGGCTGCTGCCCCTGGAG...         UKBB  OL30   \n",
       "2       GGTGTCACAGTAATGTGGAGGAGTAGAGGCACTCTGGCTTTTTGAG...         UKBB  OL27   \n",
       "3       GGTGTCACAGTAATGTGGAGGAGTAGAGGCACTCTGGCTTTTTGAG...         UKBB  OL27   \n",
       "4       GGTGTCAAGTGAGCGAGATGTCCCCGGCCACACAGTGAAGCACAAG...         UKBB  OL32   \n",
       "...                                                   ...          ...   ...   \n",
       "798059  CAGTACTGCTGGCCCCAGAAAAGCCCCTCTCCTTATACCCTAGGCC...          CRE  OL15   \n",
       "798060  CAGTACCTTGTCCCCACTTCCCATTTGGCCTCTGGCAGAGGAGGAG...          CRE  OL15   \n",
       "798061  CAGTACACCCCAGCTTCCAAAGGCCTTCTGTGACAAAGAGAGACTA...          CRE  OL15   \n",
       "798062  CAGGTTGAAGGGAATCCAAGGGCCATTTCCTTGGGAATTTCAGAAA...          CRE  OL15   \n",
       "798063  TTTTTTTTTTTTTTTTGAGATGGAGTTTCCCTCTTGTCACCCAGGC...          CRE  OL15   \n",
       "\n",
       "                  class chr  K562_log2FC  HepG2_log2FC  SKNSH_log2FC  \\\n",
       "0                   BMI   7     0.955879      0.292685     -0.780102   \n",
       "1                   SBP   6     2.059735      0.986049     -0.431290   \n",
       "2                   Plt  18     0.256503      0.266996     -0.177992   \n",
       "3                   Plt  18     0.353888      0.132264     -0.052014   \n",
       "4       Age_at_Menarche  21    -0.174600     -0.422087     -0.416915   \n",
       "...                 ...  ..          ...           ...           ...   \n",
       "798059                   18    -0.204913     -0.156933     -0.209358   \n",
       "798060                   12     1.218233      0.613623      0.569894   \n",
       "798061                    3    -0.222234     -0.338764     -0.817852   \n",
       "798062                    8    -0.194952     -0.199087      0.169847   \n",
       "798063                    2     1.064113      0.099737      0.525217   \n",
       "\n",
       "        K562_lfcSE  SKNSH_lfcSE  HepG2_lfcSE  \n",
       "0         0.559413     2.904046     1.148763  \n",
       "1         0.097276     0.089045     0.087488  \n",
       "2         0.088237     0.111069     0.100591  \n",
       "3         0.080057     0.105239     0.100011  \n",
       "4         0.184045     0.168331     0.212965  \n",
       "...            ...          ...          ...  \n",
       "798059    0.133052     0.185115     0.157279  \n",
       "798060    0.127132     0.190639     0.167222  \n",
       "798061    0.159002     0.238637     0.198187  \n",
       "798062    0.131227     0.170301     0.152272  \n",
       "798063    0.118288     0.171828     0.161234  \n",
       "\n",
       "[798064 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df = out_df.sort_values(by=['data_project'], key=lambda x: x.map({'UKBB': 0, 'GTEx': 1, 'CRE':2})).reset_index(drop=True)\n",
    "out_df = out_df.drop_duplicates('sequence', keep='first').reset_index(drop=True)\n",
    "\n",
    "out_df = out_df[out_df['chr'] != 'synth'].reset_index(drop=True)\n",
    "\n",
    "out_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88ee72ba-4f60-4e4d-ae81-4a1c9b1ec3fa",
   "metadata": {},
   "source": [
    "### Optional log2FC standard error filter\n",
    "We usually apply the log2FC standard error filter below when we analize MPRA activity and/or compare it to model predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a6217268",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IDs</th>\n",
       "      <th>sequence</th>\n",
       "      <th>data_project</th>\n",
       "      <th>OL</th>\n",
       "      <th>class</th>\n",
       "      <th>chr</th>\n",
       "      <th>K562_log2FC</th>\n",
       "      <th>HepG2_log2FC</th>\n",
       "      <th>SKNSH_log2FC</th>\n",
       "      <th>K562_lfcSE</th>\n",
       "      <th>SKNSH_lfcSE</th>\n",
       "      <th>HepG2_lfcSE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6:159961451:G:A:A:wC</td>\n",
       "      <td>GGTGTCAGAACGCCCCAGGGGGCTTTGAGGGCTGCTGCCCCTGGAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL30</td>\n",
       "      <td>SBP</td>\n",
       "      <td>6</td>\n",
       "      <td>2.059735</td>\n",
       "      <td>0.986049</td>\n",
       "      <td>-0.431290</td>\n",
       "      <td>0.097276</td>\n",
       "      <td>0.089045</td>\n",
       "      <td>0.087488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18:41953230:C:T:A:wC</td>\n",
       "      <td>GGTGTCACAGTAATGTGGAGGAGTAGAGGCACTCTGGCTTTTTGAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL27</td>\n",
       "      <td>Plt</td>\n",
       "      <td>18</td>\n",
       "      <td>0.256503</td>\n",
       "      <td>0.266996</td>\n",
       "      <td>-0.177992</td>\n",
       "      <td>0.088237</td>\n",
       "      <td>0.111069</td>\n",
       "      <td>0.100591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18:41953230:C:T:R:wC</td>\n",
       "      <td>GGTGTCACAGTAATGTGGAGGAGTAGAGGCACTCTGGCTTTTTGAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL27</td>\n",
       "      <td>Plt</td>\n",
       "      <td>18</td>\n",
       "      <td>0.353888</td>\n",
       "      <td>0.132264</td>\n",
       "      <td>-0.052014</td>\n",
       "      <td>0.080057</td>\n",
       "      <td>0.105239</td>\n",
       "      <td>0.100011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21:43635174:A:G:A:wC</td>\n",
       "      <td>GGTGTCAAGTGAGCGAGATGTCCCCGGCCACACAGTGAAGCACAAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL32</td>\n",
       "      <td>Age_at_Menarche</td>\n",
       "      <td>21</td>\n",
       "      <td>-0.174600</td>\n",
       "      <td>-0.422087</td>\n",
       "      <td>-0.416915</td>\n",
       "      <td>0.184045</td>\n",
       "      <td>0.168331</td>\n",
       "      <td>0.212965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21:43635174:A:G:R:wC</td>\n",
       "      <td>GGTGTCAAGTGAGCGAGATGTCCCCGGCCACACAGTGAAGCACAAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL32</td>\n",
       "      <td>Age_at_Menarche</td>\n",
       "      <td>21</td>\n",
       "      <td>0.081310</td>\n",
       "      <td>-0.369311</td>\n",
       "      <td>-0.377099</td>\n",
       "      <td>0.152207</td>\n",
       "      <td>0.167750</td>\n",
       "      <td>0.237189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749049</th>\n",
       "      <td>18:9125893:NA:NA</td>\n",
       "      <td>CAGTACTGCTGGCCCCAGAAAAGCCCCTCTCCTTATACCCTAGGCC...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>18</td>\n",
       "      <td>-0.204913</td>\n",
       "      <td>-0.156933</td>\n",
       "      <td>-0.209358</td>\n",
       "      <td>0.133052</td>\n",
       "      <td>0.185115</td>\n",
       "      <td>0.157279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749050</th>\n",
       "      <td>12:33905808:NA:NA</td>\n",
       "      <td>CAGTACCTTGTCCCCACTTCCCATTTGGCCTCTGGCAGAGGAGGAG...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>12</td>\n",
       "      <td>1.218233</td>\n",
       "      <td>0.613623</td>\n",
       "      <td>0.569894</td>\n",
       "      <td>0.127132</td>\n",
       "      <td>0.190639</td>\n",
       "      <td>0.167222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749051</th>\n",
       "      <td>3:128145854:NA:NA</td>\n",
       "      <td>CAGTACACCCCAGCTTCCAAAGGCCTTCTGTGACAAAGAGAGACTA...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>-0.222234</td>\n",
       "      <td>-0.338764</td>\n",
       "      <td>-0.817852</td>\n",
       "      <td>0.159002</td>\n",
       "      <td>0.238637</td>\n",
       "      <td>0.198187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749052</th>\n",
       "      <td>8:134705703:NA:NA</td>\n",
       "      <td>CAGGTTGAAGGGAATCCAAGGGCCATTTCCTTGGGAATTTCAGAAA...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "      <td>-0.194952</td>\n",
       "      <td>-0.199087</td>\n",
       "      <td>0.169847</td>\n",
       "      <td>0.131227</td>\n",
       "      <td>0.170301</td>\n",
       "      <td>0.152272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749053</th>\n",
       "      <td>2:42731164:NA:NA</td>\n",
       "      <td>TTTTTTTTTTTTTTTTGAGATGGAGTTTCCCTCTTGTCACCCAGGC...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>1.064113</td>\n",
       "      <td>0.099737</td>\n",
       "      <td>0.525217</td>\n",
       "      <td>0.118288</td>\n",
       "      <td>0.171828</td>\n",
       "      <td>0.161234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>749054 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         IDs  \\\n",
       "0       6:159961451:G:A:A:wC   \n",
       "1       18:41953230:C:T:A:wC   \n",
       "2       18:41953230:C:T:R:wC   \n",
       "3       21:43635174:A:G:A:wC   \n",
       "4       21:43635174:A:G:R:wC   \n",
       "...                      ...   \n",
       "749049      18:9125893:NA:NA   \n",
       "749050     12:33905808:NA:NA   \n",
       "749051     3:128145854:NA:NA   \n",
       "749052     8:134705703:NA:NA   \n",
       "749053      2:42731164:NA:NA   \n",
       "\n",
       "                                                 sequence data_project    OL  \\\n",
       "0       GGTGTCAGAACGCCCCAGGGGGCTTTGAGGGCTGCTGCCCCTGGAG...         UKBB  OL30   \n",
       "1       GGTGTCACAGTAATGTGGAGGAGTAGAGGCACTCTGGCTTTTTGAG...         UKBB  OL27   \n",
       "2       GGTGTCACAGTAATGTGGAGGAGTAGAGGCACTCTGGCTTTTTGAG...         UKBB  OL27   \n",
       "3       GGTGTCAAGTGAGCGAGATGTCCCCGGCCACACAGTGAAGCACAAG...         UKBB  OL32   \n",
       "4       GGTGTCAAGTGAGCGAGATGTCCCCGGCCACACAGTGAAGCACAAG...         UKBB  OL32   \n",
       "...                                                   ...          ...   ...   \n",
       "749049  CAGTACTGCTGGCCCCAGAAAAGCCCCTCTCCTTATACCCTAGGCC...          CRE  OL15   \n",
       "749050  CAGTACCTTGTCCCCACTTCCCATTTGGCCTCTGGCAGAGGAGGAG...          CRE  OL15   \n",
       "749051  CAGTACACCCCAGCTTCCAAAGGCCTTCTGTGACAAAGAGAGACTA...          CRE  OL15   \n",
       "749052  CAGGTTGAAGGGAATCCAAGGGCCATTTCCTTGGGAATTTCAGAAA...          CRE  OL15   \n",
       "749053  TTTTTTTTTTTTTTTTGAGATGGAGTTTCCCTCTTGTCACCCAGGC...          CRE  OL15   \n",
       "\n",
       "                  class chr  K562_log2FC  HepG2_log2FC  SKNSH_log2FC  \\\n",
       "0                   SBP   6     2.059735      0.986049     -0.431290   \n",
       "1                   Plt  18     0.256503      0.266996     -0.177992   \n",
       "2                   Plt  18     0.353888      0.132264     -0.052014   \n",
       "3       Age_at_Menarche  21    -0.174600     -0.422087     -0.416915   \n",
       "4       Age_at_Menarche  21     0.081310     -0.369311     -0.377099   \n",
       "...                 ...  ..          ...           ...           ...   \n",
       "749049                   18    -0.204913     -0.156933     -0.209358   \n",
       "749050                   12     1.218233      0.613623      0.569894   \n",
       "749051                    3    -0.222234     -0.338764     -0.817852   \n",
       "749052                    8    -0.194952     -0.199087      0.169847   \n",
       "749053                    2     1.064113      0.099737      0.525217   \n",
       "\n",
       "        K562_lfcSE  SKNSH_lfcSE  HepG2_lfcSE  \n",
       "0         0.097276     0.089045     0.087488  \n",
       "1         0.088237     0.111069     0.100591  \n",
       "2         0.080057     0.105239     0.100011  \n",
       "3         0.184045     0.168331     0.212965  \n",
       "4         0.152207     0.167750     0.237189  \n",
       "...            ...          ...          ...  \n",
       "749049    0.133052     0.185115     0.157279  \n",
       "749050    0.127132     0.190639     0.167222  \n",
       "749051    0.159002     0.238637     0.198187  \n",
       "749052    0.131227     0.170301     0.152272  \n",
       "749053    0.118288     0.171828     0.161234  \n",
       "\n",
       "[749054 rows x 12 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df = out_df.copy()\n",
    "stderr_columns = ['K562_lfcSE', 'HepG2_lfcSE', 'SKNSH_lfcSE']\n",
    "stderr_threshold = 1\n",
    "\n",
    "quality_filter = temp_df[stderr_columns].max(axis=1) < stderr_threshold\n",
    "temp_df = temp_df[quality_filter].reset_index(drop=True)\n",
    "temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05d213c8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
