{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "34ecaada",
   "metadata": {},
   "source": [
    "# MPRA training (for training Malinois' weights)\n",
    "\n",
    "This notebook process the MPRA results to be fed to Malinois during the training stage. The table this notebook generates was exclusively used for optimizing Malinois. Every analysis and perfomance report that uses the MPRA activity from these libraries were created using a slightly differently processed (with no DNA-count filter before averaging across libraries) table provided in the notebook \"CODA_preprocess_MPRA_validation\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "37018cd7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/sq/pxxh686s20scbcsrvq340czj6yvqp0/T/ipykernel_44940/2808823062.py:5: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display\n",
      "  from IPython.core.display import display, HTML\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>.container { width:98% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import gzip\n",
    "import os\n",
    "\n",
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:98% !important; }</style>\"))\n",
    "\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b423307",
   "metadata": {},
   "source": [
    "### Fetch the MPRA results\n",
    "\n",
    "The dataset consists of oligo libraries (OL) from the MPRA projects UKBB (OL27-33), GTEx (OL41,42), and CRE (OL15)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "17d95514",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching files:\n",
      "\n",
      "OL42_HepG2_20211010.out\n",
      "OL28_SKNSH_20210428.out\n",
      "OL41_SKNSH_20211010.out\n",
      "OL42_SKNSH_20211010.out\n",
      "OL28_HEPG2_20210428.out\n",
      "OL27_K562_20210428.out\n",
      "OL33_HEPG2_20210428.out\n",
      "OL15_HEPG2_Neon_20200904.out\n",
      "OL30_SKNSH_20210428.out\n",
      "OL33_SKNSH_20210428.out\n",
      "OL41B_HepG2_20211010.out\n",
      "OL30_HEPG2_20210428.out\n",
      "OL32_SKNSH_20210428.out\n",
      "OL33_K562_20210428.out\n",
      "OL15_K562_20200904.out\n",
      "OL31_HEPG2_20210428.out\n",
      "OL15_SKNSH_20200904.out\n",
      "OL32_HEPG2_20210428.out\n",
      "OL32_K562_20210428.out\n",
      "OL31_SKNSH_20210428.out\n",
      "OL27_SKNSH_20210428.out\n",
      "OL28_K562_20210428.out\n",
      "OL31_K562_20210428.out\n",
      "OL29_HEPG2_20210428.out\n",
      "OL27_HEPG2_20210428.out\n",
      "OL29_K562_20210428.out\n",
      "OL30_K562_20210428.out\n",
      "OL29_SKNSH_20210428.out\n",
      "OL41-42_K562_20211010.out\n"
     ]
    }
   ],
   "source": [
    "rootdir = '/training'\n",
    "\n",
    "cell_types_lower = ['k562', 'hepg2', 'sknsh']\n",
    "cell_types = ['K562', 'HepG2', 'SKNSH']\n",
    "std_cell_names = dict(zip(cell_types_lower, cell_types))\n",
    "\n",
    "UKBB_libraries = [f'OL{n}' for n in range(27, 34)]\n",
    "GTEx_libraries = ['OL41', 'OL42', 'OL41-42', 'OL41B']\n",
    "CREs_libraries = ['OL15']\n",
    "\n",
    "data_project_dict = {OL: 'UKBB' for OL in UKBB_libraries}\n",
    "data_project_dict.update({OL: 'GTEx' for OL in GTEx_libraries})\n",
    "data_project_dict.update({OL: 'CRE' for OL in CREs_libraries})\n",
    "\n",
    "print('Fetching files:\\n')\n",
    "in_df = []\n",
    "for subdir, dirs, files in os.walk(rootdir):\n",
    "    for file in files:\n",
    "        filepath = subdir + os.sep + file\n",
    "        if filepath.endswith(\".out\") and ('Counts' not in filepath):\n",
    "            library_num, cell_type = file.split('_')[:2]\n",
    "            if cell_type.lower() in cell_types_lower:\n",
    "                df_temp = pd.read_csv(filepath, sep='\\t', low_memory=False)\n",
    "                df_temp['OL'] = library_num\n",
    "                df_temp['cell_type'] = std_cell_names[cell_type.lower()]\n",
    "                df_temp['data_project'] = data_project_dict[library_num]\n",
    "                in_df.append(df_temp)\n",
    "                print(file)\n",
    "\n",
    "in_columns = ['ID', 'chr', 'project', 'ctrl_exp', 'Ctrl.Mean',\n",
    "       'Exp.Mean', 'log2FoldChange', 'lfcSE', 'OL',\n",
    "       'cell_type', 'data_project']\n",
    "\n",
    "in_df = pd.concat(in_df)[in_columns]\n",
    "\n",
    "row_filter = in_df['chr'].isna()\n",
    "chrs = []\n",
    "for ID in in_df[row_filter]['ID'].tolist():\n",
    "    if 'sample' in ID:\n",
    "        chrs.append('synth')\n",
    "    else:\n",
    "        chrs.append(ID.split(':')[0])\n",
    "        \n",
    "in_df.loc[row_filter, 'chr'] = chrs\n",
    "\n",
    "in_df.loc[in_df['project'].isna(), 'project'] = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93d59c6c",
   "metadata": {},
   "source": [
    "### Filter low-count oligos\n",
    "\n",
    "To ensure Malinois is trained only on oligos with acceptable quality, we dropped any potential noisy log2FC results before we average activities across libraries in UKBB. We imposed an oligo count threshold of at least 20, and a threshold of non-zero RNA count. For general purposes, perhaps a more suitable oligo count threhold could be at least 100 oligos. However, for the purposes of training Malinois, we did not observed any significant gains from further increasing the stringency of these thresholds. (Note that for performing analyses and figures we use a version of this table with no oligo-count threshold, only the non-zero RNA count threshold. Instead, we only filter based on an optional log2(Fold-Change) Standard Error imposed after processing the results.)\n",
    "\n",
    "Note: in order to reproduce the test set reported in the paper, one would need to set oligo_count_cutoff = 0 and filter by lfcSE after generating the table (as described in the Methods section of the paper)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "012b36a0-b55a-4b14-a4c8-d040fce5807b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>chr</th>\n",
       "      <th>project</th>\n",
       "      <th>ctrl_exp</th>\n",
       "      <th>Ctrl.Mean</th>\n",
       "      <th>Exp.Mean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>OL</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>data_project</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1000156:C:T:A:wC</td>\n",
       "      <td>1</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>90.451376</td>\n",
       "      <td>116.787452</td>\n",
       "      <td>0.370168</td>\n",
       "      <td>0.315274</td>\n",
       "      <td>OL42</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>GTEx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1000156:C:T:R:wC</td>\n",
       "      <td>1</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>81.044871</td>\n",
       "      <td>138.000723</td>\n",
       "      <td>0.764444</td>\n",
       "      <td>0.316890</td>\n",
       "      <td>OL42</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>GTEx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:10002921:T:G:A:wC</td>\n",
       "      <td>1</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>1151.172596</td>\n",
       "      <td>4035.711481</td>\n",
       "      <td>1.809318</td>\n",
       "      <td>0.082221</td>\n",
       "      <td>OL42</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>GTEx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:10002921:T:G:R:wC</td>\n",
       "      <td>1</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>974.523926</td>\n",
       "      <td>1669.391661</td>\n",
       "      <td>0.776754</td>\n",
       "      <td>0.099932</td>\n",
       "      <td>OL42</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>GTEx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:10003320:G:A:A:wC</td>\n",
       "      <td>1</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>161.422841</td>\n",
       "      <td>456.978661</td>\n",
       "      <td>1.499823</td>\n",
       "      <td>0.186002</td>\n",
       "      <td>OL42</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>GTEx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493780</th>\n",
       "      <td>X:99996102:A:G:R:wC</td>\n",
       "      <td>X</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>1138.914651</td>\n",
       "      <td>1375.044964</td>\n",
       "      <td>0.272142</td>\n",
       "      <td>0.166480</td>\n",
       "      <td>OL41-42</td>\n",
       "      <td>K562</td>\n",
       "      <td>GTEx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493781</th>\n",
       "      <td>X:99998829:C:T:A:wC</td>\n",
       "      <td>X</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>151.525843</td>\n",
       "      <td>171.747787</td>\n",
       "      <td>0.181686</td>\n",
       "      <td>0.381688</td>\n",
       "      <td>OL41-42</td>\n",
       "      <td>K562</td>\n",
       "      <td>GTEx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493782</th>\n",
       "      <td>X:99998829:C:T:R:wC</td>\n",
       "      <td>X</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>165.073711</td>\n",
       "      <td>174.544210</td>\n",
       "      <td>0.081406</td>\n",
       "      <td>0.425423</td>\n",
       "      <td>OL41-42</td>\n",
       "      <td>K562</td>\n",
       "      <td>GTEx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493783</th>\n",
       "      <td>X:99999349:G:A:A:wC</td>\n",
       "      <td>X</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>886.880586</td>\n",
       "      <td>739.964383</td>\n",
       "      <td>-0.261358</td>\n",
       "      <td>0.211378</td>\n",
       "      <td>OL41-42</td>\n",
       "      <td>K562</td>\n",
       "      <td>GTEx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493784</th>\n",
       "      <td>X:99999349:G:A:R:wC</td>\n",
       "      <td>X</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>GTEx</td>\n",
       "      <td>1106.084169</td>\n",
       "      <td>1090.048385</td>\n",
       "      <td>-0.021797</td>\n",
       "      <td>0.158862</td>\n",
       "      <td>OL41-42</td>\n",
       "      <td>K562</td>\n",
       "      <td>GTEx</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2835096 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         ID chr project ctrl_exp    Ctrl.Mean     Exp.Mean  \\\n",
       "0        1:1000156:C:T:A:wC   1    GTEx     GTEx    90.451376   116.787452   \n",
       "1        1:1000156:C:T:R:wC   1    GTEx     GTEx    81.044871   138.000723   \n",
       "2       1:10002921:T:G:A:wC   1    GTEx     GTEx  1151.172596  4035.711481   \n",
       "3       1:10002921:T:G:R:wC   1    GTEx     GTEx   974.523926  1669.391661   \n",
       "4       1:10003320:G:A:A:wC   1    GTEx     GTEx   161.422841   456.978661   \n",
       "...                     ...  ..     ...      ...          ...          ...   \n",
       "493780  X:99996102:A:G:R:wC   X    GTEx     GTEx  1138.914651  1375.044964   \n",
       "493781  X:99998829:C:T:A:wC   X    GTEx     GTEx   151.525843   171.747787   \n",
       "493782  X:99998829:C:T:R:wC   X    GTEx     GTEx   165.073711   174.544210   \n",
       "493783  X:99999349:G:A:A:wC   X    GTEx     GTEx   886.880586   739.964383   \n",
       "493784  X:99999349:G:A:R:wC   X    GTEx     GTEx  1106.084169  1090.048385   \n",
       "\n",
       "        log2FoldChange     lfcSE       OL cell_type data_project  \n",
       "0             0.370168  0.315274     OL42     HepG2         GTEx  \n",
       "1             0.764444  0.316890     OL42     HepG2         GTEx  \n",
       "2             1.809318  0.082221     OL42     HepG2         GTEx  \n",
       "3             0.776754  0.099932     OL42     HepG2         GTEx  \n",
       "4             1.499823  0.186002     OL42     HepG2         GTEx  \n",
       "...                ...       ...      ...       ...          ...  \n",
       "493780        0.272142  0.166480  OL41-42      K562         GTEx  \n",
       "493781        0.181686  0.381688  OL41-42      K562         GTEx  \n",
       "493782        0.081406  0.425423  OL41-42      K562         GTEx  \n",
       "493783       -0.261358  0.211378  OL41-42      K562         GTEx  \n",
       "493784       -0.021797  0.158862  OL41-42      K562         GTEx  \n",
       "\n",
       "[2835096 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a15ac417",
   "metadata": {},
   "outputs": [],
   "source": [
    "oligo_count_cutoff = 20\n",
    "RNA_count_cutoff = 0\n",
    "\n",
    "quality_filter = (in_df['Ctrl.Mean'] >= oligo_count_cutoff) & \\\n",
    "                         (in_df['Exp.Mean'] > RNA_count_cutoff)\n",
    "\n",
    "in_df = in_df[quality_filter].reset_index(drop=True) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec14062d",
   "metadata": {},
   "source": [
    "### Get the nucleotide sequence\n",
    "\n",
    "We load the nuecleotide sequence information from each corresponding fasta file and assign to each oligo ID its corresponding sequence. It is worth noting that there might be very rare edge cases in which the same oligo ID might differ by one position between the UKBB and GTEx databases. To our knowledge, there is only one case. Nonetheless, fetching the sequence from the corresponding OL fasta file (as opposed to just creating a set of ID-to_sequence pairs) ensures we get the correct measured sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "80ed8afc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing fasta files:\n",
      "\n",
      "OL41_reference.fasta.gz\n",
      "OL32_reference.fasta.gz\n",
      "OL31_reference.fasta.gz\n",
      "OL42_reference.fasta.gz\n",
      "OL27_reference.fasta.gz\n",
      "OL29_reference.fasta.gz\n",
      "OL15_reference.fasta.gz\n",
      "OL28_reference.fasta.gz\n",
      "OL30_reference.fasta.gz\n",
      "OL33_reference.fasta.gz\n"
     ]
    }
   ],
   "source": [
    "rootdir = '/Fastas'\n",
    "\n",
    "ignore_libraries = ['OL46']\n",
    "\n",
    "fasta_dict = {'UKBB': {}, 'GTEx': {}, 'CRE': {}}\n",
    "print('Processing fasta files:\\n')\n",
    "for subdir, dirs, files in os.walk(rootdir):\n",
    "    for file in files:\n",
    "        filepath = subdir + os.sep + file\n",
    "        if filepath.endswith(\".gz\") and ('fasta' in filepath):\n",
    "            library_num = file.split('_')[0]\n",
    "            if library_num not in ignore_libraries:\n",
    "                print(file)\n",
    "                project = data_project_dict[library_num]\n",
    "                with gzip.open(filepath, 'rt') as f:\n",
    "                    for line_str in f:\n",
    "                        if line_str[0] == '>':\n",
    "                            oligo_id = line_str.lstrip('>').rstrip('\\n')\n",
    "                            fasta_dict[project][oligo_id] = ''\n",
    "                        else:\n",
    "                            fasta_dict[project][oligo_id] += line_str.rstrip('\\n')\n",
    "\n",
    "for project in fasta_dict.keys():\n",
    "    project_fasta_dict = fasta_dict[project]\n",
    "    compressed_IDs = [ID for ID in project_fasta_dict.keys() if ';' in ID]\n",
    "    for compressed_ID in compressed_IDs:\n",
    "        stripped_ID = compressed_ID.lstrip('(').rstrip(')')\n",
    "        for single_ID in stripped_ID.split(';'):\n",
    "            project_fasta_dict[single_ID] = project_fasta_dict[compressed_ID]\n",
    "        \n",
    "in_df['sequence'] = in_df.apply(lambda x: fasta_dict[x['data_project']][x['ID']], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bee8bd2",
   "metadata": {},
   "source": [
    "### Grouping results for oligos shared across libraries in a data project\n",
    "\n",
    "This step is needed only for sequences in the UKBB oligo libraries, since these libraries share some sequences. However it does hurt to do it for the libraries in GTEx and CRE. For practical purposes, we average the log2(Fold-Change) readout of sequences shared across libraries in order to have a single sequence-to-activity pair for each sequence. For measures that can be inform us of empirical noise (such as oligo count, RNA count, and lof2FC standard error), we retrieve the \"worst\" value (min, min, max, respectively) across libraries.\n",
    "\n",
    "Note 1: For legacy reasons, we do not average shared sequences across projects since the generation of these projects was asynchronous. However, the consistency of the MPRA results probably allows for averaging across different projects as well.\n",
    "\n",
    "Note 2: We believe that many other ways of aggregating the results can achieve equal or better results for training models. For example, it could be equally useful (if not more) to show a model several sequence-to-activity pairs associated to a single sequence, as it could be a way to implicitly showing the empirical noise to the model during the training. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3f6aec13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Grouping sequences across cell types and across libraries in UKBB\n",
      "Grouping sequences across cell types and across libraries in GTEx\n",
      "Grouping sequences across cell types and across libraries in CRE\n"
     ]
    }
   ],
   "source": [
    "grouped_project_dfs = []\n",
    "for data_project in ['UKBB', 'GTEx', 'CRE']:\n",
    "    print(f'Grouping sequences across cell types and across libraries in {data_project}')\n",
    "    project_df = in_df[in_df['data_project'] == data_project].reset_index(drop=True)\n",
    "    groups = project_df.groupby(by=['sequence', 'cell_type']).agg(\n",
    "                log2FC_mean = ('log2FoldChange', 'mean'),\n",
    "                chromosome = ('chr', 'first'),\n",
    "                OL = ('OL', set),\n",
    "                classes = ('project', set),\n",
    "                IDs = ('ID', set),\n",
    "                exp_mean = ('Exp.Mean', 'min'),\n",
    "                ctrl_mean = ('Ctrl.Mean', 'min'),\n",
    "                lfcSE = ('lfcSE', 'max'))\n",
    "    grouped_project_df = groups.unstack(level=-1)\n",
    "    grouped_project_df.columns = ['_'.join(col).rstrip('_') for col in grouped_project_df.columns.values]\n",
    "    grouped_project_df.reset_index(inplace=True)\n",
    "    grouped_project_df['data_project'] = data_project\n",
    "    grouped_project_dfs.append(grouped_project_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cf55942",
   "metadata": {},
   "source": [
    "### Clean the grouped columns\n",
    "\n",
    "These lines just clean up the columns generated from stacking the groups generated in the previous cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d39bc127",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_drop_list = ['chromosome_K562', 'chromosome_SKNSH',\n",
    "               'classes_K562','classes_SKNSH',\n",
    "               'IDs_K562', 'IDs_SKNSH',\n",
    "               'OL_K562', 'OL_HepG2',\n",
    "              ]\n",
    "\n",
    "new_column_name_dict = {'sequence': 'sequence',\n",
    "                    'log2FC_mean_HepG2': 'HepG2_log2FC',\n",
    "                    'log2FC_mean_K562': 'K562_log2FC',\n",
    "                    'log2FC_mean_SKNSH': 'SKNSH_log2FC',\n",
    "                    'chromosome_HepG2': 'chr',\n",
    "                    'classes_HepG2': 'class',\n",
    "                    'IDs_HepG2': 'IDs',\n",
    "                    'OL_SKNSH': 'OL',\n",
    "                    'exp_mean_HepG2': 'exp_mean_HepG2',\n",
    "                    'exp_mean_K562': 'exp_mean_K562',\n",
    "                    'exp_mean_SKNSH': 'exp_mean_SKNSH',\n",
    "                    'ctrl_mean_HepG2': 'ctrl_mean_HepG2',\n",
    "                    'ctrl_mean_K562': 'ctrl_mean_K562',\n",
    "                    'ctrl_mean_SKNSH': 'ctrl_mean_SKNSH',\n",
    "                    'lfcSE_HepG2': 'HepG2_lfcSE',\n",
    "                    'lfcSE_K562': 'K562_lfcSE',\n",
    "                    'lfcSE_SKNSH': 'SKNSH_lfcSE',\n",
    "                    'data_project': 'data_project',\n",
    "                   }\n",
    "\n",
    "out_column_order = ['IDs', 'sequence', 'data_project', 'OL', 'class', 'chr',\n",
    "                    'K562_log2FC', 'HepG2_log2FC', 'SKNSH_log2FC',\n",
    "                    'K562_lfcSE', 'SKNSH_lfcSE', 'HepG2_lfcSE', \n",
    "#                     'ctrl_mean_K562', 'ctrl_mean_HepG2', 'ctrl_mean_SKNSH',\n",
    "#                     'exp_mean_K562', 'exp_mean_HepG2', 'exp_mean_SKNSH',\n",
    "                   ]\n",
    "\n",
    "out_df = pd.concat(grouped_project_dfs)\n",
    "\n",
    "out_df.drop(column_drop_list, axis=1, inplace=True)\n",
    "out_df.columns = [new_column_name_dict[old_name] for old_name in out_df.columns.values]\n",
    "\n",
    "null_filter = (out_df['HepG2_log2FC'].notnull()) & (out_df['K562_log2FC'].notnull()) & (out_df['SKNSH_log2FC'].notnull())\n",
    "out_df = out_df[null_filter].reset_index(drop=True)\n",
    "\n",
    "out_df['OL'] = out_df['OL'].apply(lambda x: ','.join(sorted(x)))\n",
    "out_df['class'] = out_df['class'].apply(lambda x: ','.join(sorted(x)))\n",
    "out_df['IDs'] = out_df['IDs'].apply(lambda x: ';'.join(sorted(x)))\n",
    "\n",
    "compressed_ids_filter = out_df['IDs'].str.contains(';')\n",
    "out_df.loc[compressed_ids_filter, 'IDs'] = out_df[compressed_ids_filter]['IDs'].apply(lambda x: '(' + ';'.join(x.split(';')) + ')')\n",
    "\n",
    "out_df = out_df[out_column_order]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e60a96d",
   "metadata": {},
   "source": [
    "### Drop duplicated sequences across projects\n",
    "\n",
    "In order to keep a single sequence-to-activity pair for each sequence, we drop duplicated oligos across projects. If a duplicated oligo is present in UKBB, we keep only the UKBB information. If a duplicated oligo is present in GTEx and CRE but not UKBB, we keep only the GTEx information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "481a7272",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IDs</th>\n",
       "      <th>sequence</th>\n",
       "      <th>data_project</th>\n",
       "      <th>OL</th>\n",
       "      <th>class</th>\n",
       "      <th>chr</th>\n",
       "      <th>K562_log2FC</th>\n",
       "      <th>HepG2_log2FC</th>\n",
       "      <th>SKNSH_log2FC</th>\n",
       "      <th>K562_lfcSE</th>\n",
       "      <th>SKNSH_lfcSE</th>\n",
       "      <th>HepG2_lfcSE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12:103293737:G:A:A:wC</td>\n",
       "      <td>AAAAAAAAAAAAAAAAAAAAAAAAAAAAGTGAGTAACAAAAACAAC...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL33</td>\n",
       "      <td>DVT,Asthma,Cholelithiasis</td>\n",
       "      <td>12</td>\n",
       "      <td>1.114658</td>\n",
       "      <td>0.785664</td>\n",
       "      <td>0.715749</td>\n",
       "      <td>0.113273</td>\n",
       "      <td>0.102159</td>\n",
       "      <td>0.199215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12:56396768:G:C:R:wC</td>\n",
       "      <td>GGTGTGGTGGTACATACCTGTAATCTCAGCTACTTGAGAGGCTGAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL28,OL31,OL33</td>\n",
       "      <td>Asthma,DVT,Eosino,FEV1FVC</td>\n",
       "      <td>12</td>\n",
       "      <td>-0.387188</td>\n",
       "      <td>-0.208768</td>\n",
       "      <td>-0.201981</td>\n",
       "      <td>0.291659</td>\n",
       "      <td>0.729915</td>\n",
       "      <td>0.653503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12:56396768:G:C:A:wC</td>\n",
       "      <td>GGTGTGGTGGTACATACCTGTAATCTCAGCTACTTGAGAGGCTGAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL28,OL31,OL33</td>\n",
       "      <td>Asthma,DVT,Eosino,FEV1FVC</td>\n",
       "      <td>12</td>\n",
       "      <td>-0.491576</td>\n",
       "      <td>-0.300131</td>\n",
       "      <td>-0.370913</td>\n",
       "      <td>0.474038</td>\n",
       "      <td>1.305596</td>\n",
       "      <td>1.606495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15:73041767:A:G:A:wC</td>\n",
       "      <td>GGTGTGGTGGCTCATGCCTGTAATCCCAGCACTTTGGGAGGCCGAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL32</td>\n",
       "      <td>UA</td>\n",
       "      <td>15</td>\n",
       "      <td>0.353154</td>\n",
       "      <td>-0.287985</td>\n",
       "      <td>0.527210</td>\n",
       "      <td>0.171574</td>\n",
       "      <td>0.478370</td>\n",
       "      <td>0.147259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15:73041767:A:G:R:wC</td>\n",
       "      <td>GGTGTGGTGGCTCATGCCTGTAATCCCAGCACTTTGGGAGGCCGAG...</td>\n",
       "      <td>UKBB</td>\n",
       "      <td>OL32</td>\n",
       "      <td>UA</td>\n",
       "      <td>15</td>\n",
       "      <td>0.469194</td>\n",
       "      <td>-0.132364</td>\n",
       "      <td>0.816555</td>\n",
       "      <td>0.147521</td>\n",
       "      <td>0.422527</td>\n",
       "      <td>0.161479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777930</th>\n",
       "      <td>2:70141730:NA:NA</td>\n",
       "      <td>CAGGTCCCCGGAAGTTTTCCCCGGGGACAGGTCTTGGCAACAGGTC...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.523162</td>\n",
       "      <td>2.951357</td>\n",
       "      <td>2.888325</td>\n",
       "      <td>0.100441</td>\n",
       "      <td>0.141465</td>\n",
       "      <td>0.079976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777931</th>\n",
       "      <td>10:85302269:NA:NA</td>\n",
       "      <td>CAGGTCAGGCATTTGCCTCTGCCACCAAACACTAGTTATTTCCATT...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>10</td>\n",
       "      <td>-0.304211</td>\n",
       "      <td>-0.589980</td>\n",
       "      <td>-0.654606</td>\n",
       "      <td>0.150466</td>\n",
       "      <td>0.215701</td>\n",
       "      <td>0.201285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777932</th>\n",
       "      <td>8:146012002:NA:NA</td>\n",
       "      <td>CAGGTATTCGGGAGGCTGAGACAGGAGAATCCCTTGAACTCGGGAG...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "      <td>-0.187047</td>\n",
       "      <td>-0.193959</td>\n",
       "      <td>0.404501</td>\n",
       "      <td>0.253936</td>\n",
       "      <td>0.270250</td>\n",
       "      <td>0.337266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777933</th>\n",
       "      <td>8:6652084:NA:NA</td>\n",
       "      <td>CAGTCCTGTCTGTCCTTTCCAGCAGCCCCGCAGAGGCTGAACCCCC...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>8</td>\n",
       "      <td>-0.199414</td>\n",
       "      <td>0.773116</td>\n",
       "      <td>2.511860</td>\n",
       "      <td>0.130757</td>\n",
       "      <td>0.132544</td>\n",
       "      <td>0.111419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777934</th>\n",
       "      <td>2:42731164:NA:NA</td>\n",
       "      <td>TTTTTTTTTTTTTTTTGAGATGGAGTTTCCCTCTTGTCACCCAGGC...</td>\n",
       "      <td>CRE</td>\n",
       "      <td>OL15</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>1.064113</td>\n",
       "      <td>0.099737</td>\n",
       "      <td>0.525217</td>\n",
       "      <td>0.118288</td>\n",
       "      <td>0.171828</td>\n",
       "      <td>0.161234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>777935 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          IDs  \\\n",
       "0       12:103293737:G:A:A:wC   \n",
       "1        12:56396768:G:C:R:wC   \n",
       "2        12:56396768:G:C:A:wC   \n",
       "3        15:73041767:A:G:A:wC   \n",
       "4        15:73041767:A:G:R:wC   \n",
       "...                       ...   \n",
       "777930       2:70141730:NA:NA   \n",
       "777931      10:85302269:NA:NA   \n",
       "777932      8:146012002:NA:NA   \n",
       "777933        8:6652084:NA:NA   \n",
       "777934       2:42731164:NA:NA   \n",
       "\n",
       "                                                 sequence data_project  \\\n",
       "0       AAAAAAAAAAAAAAAAAAAAAAAAAAAAGTGAGTAACAAAAACAAC...         UKBB   \n",
       "1       GGTGTGGTGGTACATACCTGTAATCTCAGCTACTTGAGAGGCTGAG...         UKBB   \n",
       "2       GGTGTGGTGGTACATACCTGTAATCTCAGCTACTTGAGAGGCTGAG...         UKBB   \n",
       "3       GGTGTGGTGGCTCATGCCTGTAATCCCAGCACTTTGGGAGGCCGAG...         UKBB   \n",
       "4       GGTGTGGTGGCTCATGCCTGTAATCCCAGCACTTTGGGAGGCCGAG...         UKBB   \n",
       "...                                                   ...          ...   \n",
       "777930  CAGGTCCCCGGAAGTTTTCCCCGGGGACAGGTCTTGGCAACAGGTC...          CRE   \n",
       "777931  CAGGTCAGGCATTTGCCTCTGCCACCAAACACTAGTTATTTCCATT...          CRE   \n",
       "777932  CAGGTATTCGGGAGGCTGAGACAGGAGAATCCCTTGAACTCGGGAG...          CRE   \n",
       "777933  CAGTCCTGTCTGTCCTTTCCAGCAGCCCCGCAGAGGCTGAACCCCC...          CRE   \n",
       "777934  TTTTTTTTTTTTTTTTGAGATGGAGTTTCCCTCTTGTCACCCAGGC...          CRE   \n",
       "\n",
       "                    OL                      class chr  K562_log2FC  \\\n",
       "0                 OL33  DVT,Asthma,Cholelithiasis  12     1.114658   \n",
       "1       OL28,OL31,OL33  Asthma,DVT,Eosino,FEV1FVC  12    -0.387188   \n",
       "2       OL28,OL31,OL33  Asthma,DVT,Eosino,FEV1FVC  12    -0.491576   \n",
       "3                 OL32                         UA  15     0.353154   \n",
       "4                 OL32                         UA  15     0.469194   \n",
       "...                ...                        ...  ..          ...   \n",
       "777930            OL15                              2     2.523162   \n",
       "777931            OL15                             10    -0.304211   \n",
       "777932            OL15                              8    -0.187047   \n",
       "777933            OL15                              8    -0.199414   \n",
       "777934            OL15                              2     1.064113   \n",
       "\n",
       "        HepG2_log2FC  SKNSH_log2FC  K562_lfcSE  SKNSH_lfcSE  HepG2_lfcSE  \n",
       "0           0.785664      0.715749    0.113273     0.102159     0.199215  \n",
       "1          -0.208768     -0.201981    0.291659     0.729915     0.653503  \n",
       "2          -0.300131     -0.370913    0.474038     1.305596     1.606495  \n",
       "3          -0.287985      0.527210    0.171574     0.478370     0.147259  \n",
       "4          -0.132364      0.816555    0.147521     0.422527     0.161479  \n",
       "...              ...           ...         ...          ...          ...  \n",
       "777930      2.951357      2.888325    0.100441     0.141465     0.079976  \n",
       "777931     -0.589980     -0.654606    0.150466     0.215701     0.201285  \n",
       "777932     -0.193959      0.404501    0.253936     0.270250     0.337266  \n",
       "777933      0.773116      2.511860    0.130757     0.132544     0.111419  \n",
       "777934      0.099737      0.525217    0.118288     0.171828     0.161234  \n",
       "\n",
       "[777935 rows x 12 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_df = out_df.sort_values(by=['data_project'], key=lambda x: x.map({'UKBB': 0, 'GTEx': 1, 'CRE':2})).reset_index(drop=True)\n",
    "out_df = out_df.drop_duplicates('sequence', keep='first').reset_index(drop=True)\n",
    "\n",
    "out_df = out_df[out_df['chr'] != 'synth'].reset_index(drop=True)\n",
    "\n",
    "out_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9691b58a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
